klippa/src/serialize.rs

```diff
@@ -0,0 +1,378 @@
+#[allow(dead_code)]
+const INVALID_OBJ_IDX: ObjIdx = ObjIdx::MAX;
+
+#[allow(dead_code)]
+#[derive(Default)]
+struct Link {
+    width: u8,
+    is_signed: bool,
+    whence: u8,
+    bias: u32,
+    position: u32,
+    objidx: ObjIdx,
+}
+
+#[derive(Default)]
+#[allow(dead_code)]
+pub(crate) struct Serializer {
+    pub(crate) start: usize,
+    pub(crate) end: usize,
+    pub(crate) head: usize,
+    pub(crate) tail: usize,
+    // TODO: zerocopy, debug_depth
+    pub(crate) errors: SerializeErrorFlags,
+
+    data: Vec<u8>,
+    object_pool: ObjectPool,
+    // index for current Object in the object_pool
+    current: usize,
+
+    packed: Vec<ObjIdx>,
+    packed_map: FnvHashMap<usize, ObjIdx>,
+}
+
+#[allow(dead_code)]
+impl Serializer {
+    pub(crate) fn new(size: u32) -> Self {
+        let buf_size = size as usize;
+        let mut this = Serializer {
+            data: vec![0; buf_size],
+            end: buf_size,
+            tail: buf_size,
+            current: NULL_IDX,
+            packed: Vec::new(),
+            ..Default::default()
+        };
+
+        this.packed.push(INVALID_OBJ_IDX);
+        this
+    }
+
+    // Embed a single Scalar type
+    pub(crate) fn embed(&mut self, obj: impl Scalar) -> Result<usize, SerializeErrorFlags> {
+        let raw = obj.to_raw();
+        let bytes = raw.as_ref();
+        let size = bytes.len();
+
+        let ret = self.allocate_size(size)?;
+        self.data
+            .get_mut(ret..ret + size)
+            .unwrap()
+            .copy_from_slice(bytes);
+        Ok(ret)
+    }
+
+    // Allocate size
+    pub(crate) fn allocate_size(&mut self, size: usize) -> Result<usize, SerializeErrorFlags> {
+        if self.in_error() {
+            return Err(self.errors);
+        }
+
+        if size > u32::MAX as usize || self.tail - self.head < size {
+            return Err(self.err(SerializeErrorFlags::SERIALIZE_ERROR_OUT_OF_ROOM));
+        }
+
+        //TODO: add support for clear?
+        let ret = self.head;
+        self.head += size;
+        Ok(ret)
+    }
+
+    #[inline]
+    pub(crate) fn successful(&self) -> bool {
+        !self.errors
+    }
+
+    #[inline]
+    pub(crate) fn in_error(&self) -> bool {
+        !!self.errors
+    }
+
+    #[inline]
+    pub(crate) fn err(&mut self, error_type: SerializeErrorFlags) -> SerializeErrorFlags {
+        self.errors |= error_type;
+        self.errors
+    }
+
+    pub(crate) fn push(&mut self) -> Result<(), SerializeErrorFlags> {
+        if self.in_error() {
+            return Err(self.errors);
+        }
+
+        let obj_idx = self.object_pool.alloc();
+        let Some(obj) = self.object_pool.get_obj(obj_idx) else {
+            return Err(self.err(SerializeErrorFlags::SERIALIZE_ERROR_OTHER));
+        };
+
+        obj.head = self.head;
+        obj.tail = self.tail;
+        obj.next_obj = self.current;
+        self.current = obj_idx;
+
+        Ok(())
+    }
+
+    pub(crate) fn copy_bytes(&self) -> Vec<u8> {
```

Comment:
Since I don't believe we ever need multiple copies of the serialize result, you could make this more efficient by having it pass ownership of the internal vector (consuming the serialize context) instead of creating a copy. You can use https://doc.rust-lang.org/std/primitive.slice.html#method.copy_within to shift the tail bytes into the correct place then shrink the internal buffer as needed.